heat/cluster/spectral.py

```diff
@@ -0,0 +1,309 @@
+        self._cluster_centers = None
+        self._labels = None
+
+    @property
+    def components_(self):
+        """
+        Returns
+        -------
+        ht.DNDarray, shape = (n_samples, n_lanzcos):
+            Eigenvectors of the graph Laplacian
+        """
+        return self._components
+
+    @property
+    def labels_(self):
+        """
+        Returns
+        -------
+        ht.DNDarray, shape=(n_points):
+            Labels of each point.
+        """
+        return self._labels
+
+    def fit(self, X):
+        """
+        Computes the low-dim representation by calculation of eigenspectrum (eigenvalues and eigenvectors) of the graph laplacian from the similarity matrix and fits the eigenvectors that correspond to the k lowest eigenvalues with a seperate clustering algorithm (currently only kemans is supported)
+        Similarity metrics for adjacency calculations are supported via spatial.distance. The eigenvalues and eigenvectors are computed by reducing the Laplacian via lanczos iterations and using the torch eigenvalue solver on this smaller matrix. If other eigenvalue decompostion methods are supported, this will be expanded.
+
+
+        Parameters
+        ----------
+        X : ht.DNDarray, shape=(n_samples, n_features)
+            Training instances to cluster.
+        """
+        # input sanitation
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("input needs to be a ht.DNDarray, but was {}".format(type(X)))
+        if X.split is not None and X.split != 0:
+            raise NotImplementedError("Not implemented for other splitting-axes")
+        # 1. Calculation of Adjacency Matrix
+        if self.metric == "rbf":
+            sig = math.sqrt(1 / (2 * self.gamma))
+            S = ht.spatial.rbf(X, sigma=sig, quadratic_expansion=True)
+
+        elif self.metric == "euclidean":
+            S = ht.spatial.cdist(X)
+        else:
+            raise NotImplementedError("Other kernels currently not implemented")
+
+        # 2. Calculation of Laplacian
+        if self.laplacian == "eNeighborhood":
+            if self.epsilon[1] == "upper":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", upper=self.epsilon[0])
+            elif self.epsilon[1] == "lower":
+                L = laplacian(S, norm=self.normalize, mode="eNeighbour", lower=self.epsilon[0])
+            else:
+                raise ValueError(
+                    "Boundary needs to be 'upper' or 'lower' and threshold needs to be set, if laplacian = eNeighborhood"
+                )
+
+        elif self.laplacian == "fully_connected":
+            L = laplacian(S, norm=self.normalize, mode="fc")
+        else:
+            raise NotImplementedError("Other approaches currently not implemented")
+
+        # 3. Eigenvalue and -vector calculation via Lanczos Algorithm
+        v0 = ht.ones((L.shape[0],), dtype=L.dtype, split=L.split, device=L.device) / math.sqrt(
```

Comment:
ht.sqrt?

Reply:
for a simple float computation? seems a bit much of an overhead...

Reply:
well scalars should be equally fast